crates/miden-objects/src/account/storage/header.rs

```diff
@@ -208,8 +277,21 @@ impl Serializable for AccountStorageHeader {
 impl Deserializable for AccountStorageHeader {
     fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
         let len = source.read_u8()?;
-        let slots = source.read_many(len as usize)?;
-        Self::new(slots).map_err(|err| DeserializationError::InvalidValue(err.to_string()))
+        let slots: Vec<StorageSlotHeader<'static>> = source.read_many(len as usize)?;
+
+        if slots.len() > AccountStorage::MAX_NUM_STORAGE_SLOTS {
+            return Err(DeserializationError::InvalidValue(
+                AccountError::StorageTooManySlots(slots.len() as u64).to_string(),
+            ));
+        }
+
+        if !slots.is_sorted_by_key(|slot| slot.id()) {
+            return Err(DeserializationError::InvalidValue(
+                AccountError::UnsortedStorageSlots.to_string(),
+            ));
+        }
+
+        Ok(Self { slots })
```

Comment:
It would be good not do duplicate this logic to check sorting, etc. It'd be best to keep calling `Self::new`.

I think we can change `AccountStorageHeader::new` to take `Vec<StorageSlotHeader<'static>>` and then it should be easy to deduplicate this logic.